pkg/storegateway/series_chunks.go

```diff
@@ -417,6 +422,183 @@ func (c *loadingSeriesChunksSetIterator) Err() error {
 	return c.err
 }
 
+//nolint:unused // dead code while we are working on PR 3968
+type partialSeriesChunksSet struct {
+	//nolint:unused // dead code while we are working on PR 3968
+	refsRanges []seriesChunkRefsRange
+	rawRanges  [][]byte
+	//nolint:unused // dead code while we are working on PR 3968
+	parsedChunks []storepb.AggrChunk
+}
+
+//nolint:unused // dead code while we are working on PR 3968
+func newPartialSeries(nextUnloaded seriesChunkRefsSet, getPooledChunks func(size int) []storepb.AggrChunk) []partialSeriesChunksSet {
+	partialSeries := make([]partialSeriesChunksSet, len(nextUnloaded.series))
+
+	for i, s := range nextUnloaded.series {
+		numRanges := len(s.chunksRanges)
+		partialSeries[i].refsRanges = s.chunksRanges
+		partialSeries[i].rawRanges = make([][]byte, numRanges)
+		chunksCount := 0
+		for _, g := range s.chunksRanges {
+			chunksCount += len(g.refs)
+		}
+		partialSeries[i].parsedChunks = getPooledChunks(chunksCount)
+	}
+	return partialSeries
+}
+
+//nolint:unused // dead code while we are working on PR 3968
+type underfetchedChunksRangeIdx struct {
+	blockID  ulid.ULID
+	rangeIdx int
+	parsed   []storepb.AggrChunk
+}
+
+// parse tries to parse the ranges in the partial set with the bytes it has in rawRanges.
+// If the bytes for any of the ranges aren't enough, then parse will return an underfetchedChunksRangeIdx
+// and will correct the length of all ranges which had a understimated length.
+// Currently, parse will only correct the length of the last chunk, since this is the only chunk
+// whose size we estimate.
+// parse will also truncate the bytes in rawRanges in case there are extra unnecessary bytes there.
+// parse will return an error when the bytes in rawRanges were underfetched before the last chunk or
+// if the data in rawRanges is invalid.
+//
+//nolint:unused // dead code while we are working on PR 3968
+func (s partialSeriesChunksSet) parse() ([]underfetchedChunksRangeIdx, error) {
```

Comment:
This looks testable, given `partialSeriesChunksSet` is self contained and doesn't rely on any external dependancy.